server/src/main/java/org/elasticsearch/transport/ConnectionManager.java

```diff
@@ -75,66 +82,119 @@ public void removeListener(TransportConnectionListener listener) {
-                    return;
-                }
-                boolean success = false;
-                try {
-                    connection = internalOpenConnection(node, resolvedProfile);
-                    connectionValidator.accept(connection, resolvedProfile);
-                    // we acquire a connection lock, so no way there is an existing connection
-                    connectedNodes.put(node, connection);
-                    if (logger.isDebugEnabled()) {
-                        logger.debug("connected to node [{}]", node);
-                    }
+
+        if (connectingRefCounter.tryIncRef() == false) {
+            listener.onFailure(new IllegalStateException("connection manager is closed"));
+            return;
+        }
+
+        try (Releasable lock = connectionLock.acquire(node.getId())) {
+            Transport.Connection connection = connectedNodes.get(node);
+            if (connection != null) {
+                assert connectingNodes.containsKey(node) == false;
+                lock.close();
+                connectingRefCounter.decRef();
+                listener.onResponse(null);
+                return;
+            }
+
+            final List<ActionListener<Void>> connectionListeners = connectingNodes.computeIfAbsent(node, n -> new ArrayList());
+            connectionListeners.add(listener);
+            if (connectionListeners.size() > 1) {
+                // wait on previous entry to complete connection attempt
+                connectingRefCounter.decRef();
+                return;
+            }
+        }
+
+        final RunOnce releaseOnce = new RunOnce(connectingRefCounter::decRef);
+
+        internalOpenConnection(node, resolvedProfile, ActionListener.wrap(conn -> {
+            connectionValidator.validate(conn, resolvedProfile, ActionListener.wrap(
+                ignored -> {
+                    assert Transports.assertNotTransportThread("connection validator success");
+                    boolean success = false;
+                    List<ActionListener<Void>> listeners = null;
                     try {
-                        connectionListener.onNodeConnected(node);
+                        // we acquire a connection lock, so no way there is an existing connection
+                        try (Releasable ignored2 = connectionLock.acquire(node.getId())) {
+                            connectedNodes.put(node, conn);
+                            if (logger.isDebugEnabled()) {
+                                logger.debug("connected to node [{}]", node);
+                            }
+                            try {
+                                connectionListener.onNodeConnected(node);
+                            } finally {
+                                final Transport.Connection finalConnection = conn;
+                                conn.addCloseListener(ActionListener.wrap(() -> {
+                                    logger.info("close listener called for node {}", node);
```

Comment:
I don't totally understand the value of this logging message. Especially since there might be multiple and other close listeners attached to this connection. This message seems to imply THIS is the close listener. It would make more sense to me if the message at lease reflected what this listener is going (deregistering the node and notify that it has disconnected).

Reply:
this was a leftover from a debugging session :)

Reply:
turned this into trace log at 3a0fc08